src/avram/bulk_upsert.cr

```diff
@@ -0,0 +1,82 @@
+class Avram::BulkUpsert(T)
+  def initialize(@records : Array(T), @column_names : Array(Symbol))
+    @records = set_timestamps(records)
+  end
+
+  def statement
+    <<-SQL
+    INSERT INTO #{table}(#{fields})
+      (select * from unnest(#{value_placeholders}))
+      ON CONFLICT (#{conflicts}) DO UPDATE SET #{updates}
+      RETURNING #{returning}
+    SQL
+  end
+
+  def args
+    @records.map do |record|
+      record.changed_attributes.map(&.value)
+    end.transpose
+  end
+
+  private def conflicts
+    @column_names.join(", ")
+  end
+
+  private def set_timestamps(collection)
+    collection.map do |record|
+      record.created_at.value ||= Time.utc if record.responds_to?(:created_at)
+      record.updated_at.value = Time.utc if record.responds_to?(:updated_at)
+      record
+    end
+  end
+
+  private def table
+    @records.first.table_name
+  end
+
+  private def updates
+    update_keys = changed_attributes.flat_map(&.name)
+    (update_keys - [:created_at]).map do |column|
+      "#{column}=EXCLUDED.#{column}"
+    end.join(", ")
+  end
+
+  private def returning
+    T.column_names.join(", ")
+  end
+
+  private def changed_attributes
+    @records.first.changed_attributes
```

Comment:
I don't know how arrays work in crystal, but `@records.first` is used _a lot_ -- might be worth memoizing

I suggest the name `sample`